metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -568,45 +576,62 @@ void cancel() {
         }
 
         void reschedule(long delayNs) {
-            ControlEvent event = new ControlEvent(GENERATE_SNAPSHOT, this);
+            ControllerEvent event = new ControllerEvent(GENERATE_SNAPSHOT, this);
             queue.scheduleDeferred(event.name,
                 new EarliestDeadlineFunction(time.nanoseconds() + delayNs), event);
         }
 
+        void handleSnapshotFinished(Optional<Exception> exception) {
+            if (exception.isPresent()) {
+                log.error("Error while generating snapshot {}", generator.lastContainedLogOffset(), exception.get());
```

Comment:
If there is a snapshot failure, will we wait for the next interval before retrying? I wonder if we need a metric or something to get marked when this happens?

Reply:
> If there is a snapshot failure, will we wait for the next interval before retrying?
Yeah. The snapshot counters `committedBytesSinceLastSnapshot` and `oldestCommittedLogOnlyAppendTimestamp` are updated assuming that snapshot generation cannot fail.

We could add another reason ("snapshot failure") that triggers a snapshot immediately irrespective of the counters. I am concerned that this event/task my starve other controller events. Maybe it is better to just rely on timed snapshot and NoOpRecord to trigger another snapshot in the case of failures.

> If there is a snapshot failure, will we wait for the next interval before retrying?

Let me check the KIPs but I don't think we define this metric. I can add it and update one of the KIPs if we agree to it here.

Reply:
How about this for the controller:
`kafka.controller:type=KafkaController,name=MetadataSnapshotGenerationErrors`
Incremented anytime the controller fails to generate a snapshot. Reset to zero anytime the controller restarts or a snapshot is successfully generated.

And this for the brokers:
kafka.server:type=broker-metadata-metrics,name=snapshot-generation-errors
Incremented anytime the broker fails to generate a snapshot. Reset to zero anytime the broker restarts or a snapshot is successfully generated.

@hachikuji do you mind if I implement this in a future PR? I created this issue: https://issues.apache.org/jira/browse/KAFKA-14403

Reply:
Doing it separately sounds fine.

Reply:
>  kafka.controller:type=KafkaController,name=MetadataSnapshotGenerationErrors

I am concerned that with a counter that resets at restart, we might have a ever so small likelihood of never incrementing the metric even with the log moving if we e.g. keep crashing while generating a snapshot. Ditto for the broker.

What do you think about having a `kafka.controller:type=KafkaController,name=LastSnapshotTime` metric. 
This metric would just emit the timestamp (linux epoch) of when the last snapshot was successfully written.

Reply:
Okay. Let's move this discussion to the DISCUSSION thread in Kafka dev. I'll send a message next week.